### PR TITLE
feat: Added Resizable Wrapper for Ant Design Table

### PR DESCRIPTION
**Pull request by csudani7**
Close: https://github.com/SigNoz/signoz/issues/1792

### CODE REVIEW

File: frontend/README.md [hunk 1]
```diff
@@ -23,15 +24,16 @@ docker-compose up
 ```
 
 ## Without Docker
+
 Follow the steps below
 
-1. ```git clone https://github.com/SigNoz/signoz.git && cd signoz/frontend```
-1. change baseURL to ```<test environment URL>``` in file ```src/constants/env.ts```
```
Comment (palashgdev) [inline frontend/README.md:29]: can you please remove this unnecessary change

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/README.md (hunk 1 above)
Comment (csudani7) [reply frontend/README.md]: @palashgdev - Removed unnecessary changes & synced with upstream develop branch.

File: frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx:1]: can you please remove the eslint disabling of the rule

File: frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+/* eslint-disable react/jsx-props-no-spreading */
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx:2]: same this one also

File: frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import React from 'react';
+import { Resizable, ResizeCallbackData } from 'react-resizable';
+
+// # Styles import
+import { SpanStyle } from './styles';
+
+function ResizableHeader(
+	props: React.HTMLAttributes<any> & {
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx:10]: can you please make a named interface here

File: frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx:1]: same this one

File: frontend/src/container/ListAlertRules/ListAlert.tsx [hunk 6]
```diff
@@ -67,6 +72,7 @@ function ListAlert({ allAlertRules, refetch }: ListAlertProps): JSX.Element {
 		{
 			title: 'Alert Name',
 			dataIndex: 'alert',
+			width: 100,
```
Comment (palashgdev) [inline frontend/src/container/ListAlertRules/ListAlert.tsx:75]: do need to explicitly add the width in the initial rendering ?

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/container/ListAlertRules/ListAlert.tsx (hunk 6 above)
Comment (csudani7) [reply frontend/src/container/ListAlertRules/ListAlert.tsx:75]: @palashgdev 
This is for resize header after initial rendering width. This is default width of header later on we can resize according to our conviction.

File: frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx (hunk 2 above)
Comment (csudani7) [reply frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx]: Done

File: frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx (hunk 4 above)
Comment (csudani7) [reply frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx]: Created. Done

**Comment by ankitnayan** [comment]
It overall looks good. But there is high sensitivity on cursor drag. If we can have a quick fix for that then it would be awesome otherwise LGTM. Sharing a loom for the same
https://www.loom.com/share/129a607749bd4a359d3227bc59223ff1

File: frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	): void => {
+		const newColumns = [...columnsData];
+		newColumns[index] = {
+			...newColumns[index],
+			width: size.width,
+		};
+		setColumns(newColumns);
+	};
+
+	const mergeColumns = columnsData.map((col, index) => ({
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx:21]: useMemo

File: frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+		...col,
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx:23]: can we define the type here ?

File: frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+			width: column.width,
+			onResize: handleResize(index),
+		}),
+	}));
+
+	return <> {React.cloneElement(children, { columns: mergeColumns })}</>;
+}
+
+interface ResizeTableWrapperProps {
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx:34]: same here

File: frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+function ResizeTableWrapper(props: ResizeTableWrapperProps): JSX.Element {
+	const { columns, children } = props;
+	const [columnsData, setColumns] = useState<ColumnsType>(columns);
+
+	const handleResize = (index: number) => (
+		_e: React.SyntheticEvent<Element>,
+		{ size }: ResizeCallbackData,
+	): void => {
+		const newColumns = [...columnsData];
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx:13]: would it possible to move the logic in utils ?

File: frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import { ColumnsType } from 'antd/lib/table';
+import React, { useState } from 'react';
+import { ResizeCallbackData } from 'react-resizable';
+
+function ResizeTableWrapper(props: ResizeTableWrapperProps): JSX.Element {
+	const { columns, children } = props;
+	const [columnsData, setColumns] = useState<ColumnsType>(columns);
+
+	const handleResize = (index: number) => (
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx:9]: also wrap this function in usecallback

File: frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+		// eslint-disable-next-line react/jsx-props-no-spreading
+		return <th {...restProps} />;
+	}
+
+	return (
+		<Resizable
+			width={width}
+			height={0}
+			handle={
+				<SpanStyle
```
Comment (palashgdev) [inline frontend/src/components/ResizeTableWrapper/ResizableHeader.tsx:20]: useMemo

File: frontend/package.json [hunk 13]
```diff
@@ -137,6 +137,7 @@
 		"@types/react-dom": "18.0.10",
 		"@types/react-grid-layout": "^1.1.2",
 		"@types/react-redux": "^7.1.11",
+		"@types/react-resizable": "^3.0.3",
```
Comment (palashgdev) [inline frontend/package.json:140]:
```suggestion
		"@types/react-resizable": "3.0.3",
```

File: frontend/package.json [hunk 14]
```diff
@@ -175,6 +176,7 @@
 		"portfinder-sync": "^0.0.2",
 		"prettier": "2.2.1",
 		"react-hot-loader": "^4.13.0",
+		"react-resizable": "^3.0.4",
```
Comment (palashgdev) [inline frontend/package.json:179]:
```suggestion
		"react-resizable": "3.0.4",
```

File: frontend/src/container/ListOfDashboard/index.tsx [hunk 15]
```diff
@@ -271,19 +281,21 @@ function ListOfAllDashboard(): JSX.Element {
...
+					<Table
+						pagination={{
+							pageSize: 9,
+							defaultPageSize: 9,
+						}}
+						showHeader
+						bordered
+						sticky
+						loading={loading}
+						components={{ header: { cell: ResizableHeader } }}
```
Comment (palashgdev) [inline frontend/src/container/ListOfDashboard/index.tsx:294]: can wrap this object in memo

File: frontend/src/container/ListOfDashboard/index.tsx [hunk 16]
```diff
@@ -271,19 +281,21 @@ function ListOfAllDashboard(): JSX.Element {
...
-					bordered
-					sticky
-					loading={loading}
-					columns={columns}
-					dataSource={data}
-					showSorterTooltip
-				/>
+				<ResizeTableWrapper columns={columns}>
+					<Table
+						pagination={{
```
Comment (palashgdev) [inline frontend/src/container/ListOfDashboard/index.tsx:286]: as this object is constant please move this one config

File: frontend/src/container/MetricsTable/index.tsx [hunk 17]
```diff
@@ -141,12 +149,14 @@ function Metrics(): JSX.Element {
...
-				dataSource={services}
-				columns={columns}
-				rowKey="serviceName"
-			/>
+			<ResizeTableWrapper columns={columns}>
+				<Table
+					loading={loading}
+					dataSource={services}
+					components={{ header: { cell: ResizableHeader } }}
+					rowKey="serviceName"
```
Comment (palashgdev) [inline frontend/src/container/MetricsTable/index.tsx:157]: memo

File: frontend/src/container/NewDashboard/DashboardSettings/Variables/index.tsx [hunk 18]
```diff
@@ -159,7 +166,12 @@ function VariablesSetting({
 							<PlusOutlined /> New Variables
 						</Button>
 					</Row>
-					<Table columns={columns} dataSource={variablesTableData} />
+					<ResizeTableWrapper columns={columns}>
+						<Table
+							components={{ header: { cell: ResizableHeader } }}
```
Comment (palashgdev) [inline frontend/src/container/NewDashboard/DashboardSettings/Variables/index.tsx:171]: same here

File: frontend/src/container/OrganizationSettings/AuthDomains/index.tsx [hunk 19]
```diff
@@ -266,12 +274,14 @@ function AuthDomains(): JSX.Element {
...
 				</Modal>
-				<Table
-					rowKey={(record: AuthDomain): string => record.name + v4()}
-					dataSource={!SSOFlag ? notEntripriseData : []}
-					columns={columns}
-					tableLayout="fixed"
-				/>
+				<ResizeTableWrapper columns={columns}>
+					<Table
+						rowKey={(record: AuthDomain): string => record.name + v4()}
```
Comment (palashgdev) [inline frontend/src/container/OrganizationSettings/AuthDomains/index.tsx:279]: on re render this will create a new function or objects

File: frontend/src/container/OrganizationSettings/AuthDomains/index.tsx [hunk 20]
```diff
@@ -315,13 +325,15 @@ function AuthDomains(): JSX.Element {
...
-					loading={isLoading}
-					columns={columns}
-					tableLayout="fixed"
-					rowKey={(record: AuthDomain): string => record.name + v4()}
-				/>
+				<ResizeTableWrapper columns={columns}>
+					<Table
+						dataSource={tableData}
+						loading={isLoading}
+						components={{ header: { cell: ResizableHeader } }}
```
Comment (palashgdev) [inline frontend/src/container/OrganizationSettings/AuthDomains/index.tsx:332]: same here

**Review by palashgdev — CHANGES_REQUESTED** [review]

**Comment by palashgdev** [comment]
can you please confirm the packages that we adding might be the issue ?

Screenshot 2023-01-17 at 1 21 23 PM

**Comment by csudani7** [comment]
> can you please confirm the packages that we adding might be the issue ?
> 
> Screenshot 2023-01-17 at 1 21 23 PM

@palashgdev This package have issue with yarn.
When I did yarn install on develop its working.
same with my branch its not working.
once I removed react-resizable. Then its working.
PS: It is working in any case with **npm**

File: frontend/src/container/ListOfDashboard/index.tsx (hunk 16 above)
Comment (csudani7) [reply frontend/src/container/ListOfDashboard/index.tsx:286]: @palashgdev Pagination calculation is not a constant, its different for across component.

File: frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx (hunk 10 above)
Comment (csudani7) [reply frontend/src/components/ResizeTableWrapper/ResizeTableWrapper.tsx]: It is wrapped in useCallback & wrapper file so, its fine to have it on file.

File: frontend/src/container/ListOfDashboard/index.tsx (hunk 15 above)
Comment (csudani7) [reply frontend/src/container/ListOfDashboard/index.tsx:294]: for components? @palashgdev

File: frontend/src/container/OrganizationSettings/AuthDomains/index.tsx (hunk 19 above)
Comment (csudani7) [reply frontend/src/container/OrganizationSettings/AuthDomains/index.tsx:277]: it will return a unique rowKey with function.

**Review by palashgdev — APPROVED** [review]
Create a single instance of Table as components={{ header: { cell: ResizableHeader } }} is used every where
